### ISSUE TITLE
lineinfile "regexp" ignored when using "insertafter"

### ISSUE THREAD

**Issue by jsteel44** [ansible/ansible#36156]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
lineinfile

##### ANSIBLE VERSION
```
ansible 2.6.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/jsteel/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Jan  5 2018, 10:41:29) [GCC 7.2.1 20171224]
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
N/A

##### SUMMARY
Recent change #28721 breaks my config where I use both regexp AND insertafter; the regexp doesn't seem to match any more, so on every subsequent run it inserts duplicate lines.

##### STEPS TO REPRODUCE

```yaml
- name: my.cnf options
  lineinfile:
    dest: "{{ mariadb_config }}"
    regexp: "^{{ item.option }}="
    line: "{{ item.option }}={{ item.value }}"
    insertafter: '^\[{{ item.section }}\]'
  with_items:
     - { section: mysqld, option: innodb_log_file_size, value: "{{ innodb_log_file_size }}" }
     - { section: mysqld, option: default-storage-engine, value: innodb }
     - { section: mysqld, option: max_allowed_packet, value: 128M }
```

##### EXPECTED RESULTS

On subsequent runs, nothing should change.

##### ACTUAL RESULTS

A duplicate line is inserted for each item, for eg:

```
--- before: /etc/my.cnf (content)
+++ after: /etc/my.cnf (content)
@@ -1,4 +1,5 @@
 [mysqld]
+default-storage-engine=innodb
 skip-name-resolve
 max_allowed_packet=128M
 default-storage-engine=innodb
```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/modules/files/lineinfile.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/files/lineinfile.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @ahtik @dhozac
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by gundalow** [comment]
Looks like https://github.com/ansible/ansible/pull/33393 went into `stable-2.4`, `stable-2.5` and `devel`, so the fix will need to go there as well.

Would be good to an integration test for this

**Comment by samdoran** [comment]
@jsteel44 Thanks for opening this issue. I'll work on getting a fix in shortly.

**Comment by samdoran** [comment]
@jsteel44 That took me a lot longer than anticipated. Please test #36474 as soon as you can since we have releases going out soon. Thanks!

**Comment by jsteel44** [comment]
Looks good thanks!

**Comment by samdoran** [comment]
@jsteel44 Thank you for verifying. This fix will be in Ansible `2.4.4` and Ansible `2.5.0.b2`.

**Comment by muffl0n** [comment]
Sorry to resurrect this old issue but I'm still experiencing this behaviour when using "insertafter":
```
    - name: allow access for munin hosts
      lineinfile:
        dest=/etc/munin/munin-node.conf
        insertafter="^allow"
        line="allow ^{{ item.replace('.', '\.') }}$"
      with_items:
        - 127.0.0.2
        - 127.0.0.3
```
The lines get inserted over and over again. If I remove "insertafter" everything works fine. 

The fix should be in 2.5.0b2, right?
```
-> % ansible --version
ansible 2.5.0b2 (detached HEAD 62139bb828) last updated 2018/04/10 08:51:28 (GMT +200)
  config file = /home/svs/workspace.intellij/ansible/role-munin_node/ansible.cfg
  configured module search path = ['/home/svs/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /tmp/ansible/lib/ansible
  executable location = /tmp/ansible/bin/ansible
  python version = 3.6.4 (default, Jan  5 2018, 02:35:40) [GCC 7.2.1 20171224] ```

**Comment by muffl0n** [comment]
Btw, same behaviour in devel:
```ansible 2.6.0 (devel 361437b042) last updated 2018/04/10 08:30:33 (GMT +200)```

**Comment by ilyaevseev** [comment]
The same bug in ansible 2.7.7

**Comment by muffl0n** [comment]
@samdoran, could you please reopen this bug? Or should we open a new one?

**Comment by samdoran** [comment]
It seems that in fixing this bug I changed behavior which was then fixed by https://github.com/ansible/ansible/pull/49409, which was backported and released in v2.6.12 and v2.7.6.

@muffl0n Please test with newer versions of Ansible and open a new issue with a simple reproducer if it is still present.

@ilyaevseev I am unable to duplicate this issue in Ansible 2.7.7 with the following playbook:
```yaml
- name: Test lineinfile
  hosts: localhost
  connection: local
  become: no
  gather_facts: no

  tasks:
    - name: allow access for munin hosts
      lineinfile:
        dest: zzz.txt
        insertafter: "^allow"
        line: "allow ^{{ item.replace('.', '.') }}$"
      with_items:
        - 127.0.0.2
        - 127.0.0.3
```

```
> ansible --version
ansible 2.7.7 (detached HEAD 11a2bafa21) last updated 2019/03/04 10:59:09 (GMT -400)
  config file = /Users/sdoran/.ansible.cfg
  configured module search path = ['/Users/sdoran/Source/ansible/library']
  ansible python module location = /Users/sdoran/Source/ansible/lib/ansible
  executable location = /Users/sdoran/Source/ansible/bin/ansible
  python version = 3.7.2 (default, Feb 26 2019, 10:16:40) [Clang 10.0.0 (clang-1000.10.44.4)]

> cat zzz.txt
allow ^127.0.0.2$
allow ^127.0.0.3$

> ansible-playbook test-lineinfile.yml

PLAY [Test lineinfile] *****************************************************************************************************************

TASK [allow access for munin hosts] ****************************************************************************************************
ok: [localhost] => (item=127.0.0.2)
ok: [localhost] => (item=127.0.0.3)

PLAY RECAP *****************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0
```

It also seems to work properly in `devel`:
```
> ansible --version
aansible 2.8.0.dev0
  config file = /Users/sdoran/Projects/Vagrant/Ansible Lab/ansible.cfg
  configured module search path = ['/Users/sdoran/Source/ansible/library']
  ansible python module location = /Users/sdoran/Source/ansible/lib/ansible
  executable location = /Users/sdoran/Source/ansible/bin/ansible
  python version = 3.7.2 (default, Feb 26 2019, 10:16:40) [Clang 10.0.0 (clang-1000.10.44.4)]

> ansible-playbook test-lineinfile.yml

PLAY [Test lineinfile] *****************************************************************************************************************

TASK [allow access for munin hosts] ****************************************************************************************************
ok: [localhost] => (item=127.0.0.2)
ok: [localhost] => (item=127.0.0.3)

PLAY RECAP *****************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

**Comment by muffl0n** [comment]
You are right! Works for me with 2.7.6 and 2.7.8.

Thanks for the quick reply!

**Comment by ilyaevseev** [comment]
> @ilyaevseev I am unable to duplicate this issue in Ansible 2.7.7 with the following playbook:

Yes, my mistake. Duplicated inserts are caused by "firstmatch: yes", not by regexp.
